SeQuant/core/index.hpp

```diff
@@ -79,62 +110,173 @@ class Index : public Taggable {
+  /// space this object belongs
+  /// @param proto_indices labels of proto indices (all must be unique,
+  /// i.e. duplicates are not allowed)
+  /// @param symmetric_proto_indices if true, proto_indices can be permuted at
+  /// will and will always be sorted
+  Index(const IndexSpace &space, IndexList proto_indices,
+        bool symmetric_proto_indices = true)
+      : Index(space) {
+    proto_indices_ = std::move(proto_indices);
+    symmetric_proto_indices_ = symmetric_proto_indices;
+    canonicalize_proto_indices();
+    check_for_duplicate_proto_indices();
+  }
+
+  /// @param space (a const ref to) the IndexSpace object that specifies to this
+  /// space this object belongs
+  /// @param ord the index ordinal
+  /// @param proto_indices labels of proto indices (all must be unique,
+  /// i.e. duplicates are not allowed)
+  /// @param symmetric_proto_indices if true, proto_indices can be permuted at
+  /// will and will always be sorted
+  Index(const IndexSpace &space, meta::integral auto ord,
         container::vector<Index> proto_indices,
         bool symmetric_proto_indices = true)
-      : label_(to_wstring(std::forward<String>(label))),
-        space_(space),
-        proto_indices_(std::move(proto_indices)),
-        symmetric_proto_indices_(symmetric_proto_indices) {
+      : Index(space, ord) {
+    proto_indices_ = std::move(proto_indices);
+    symmetric_proto_indices_ = symmetric_proto_indices;
+    canonicalize_proto_indices();
+    check_for_duplicate_proto_indices();
+  }
+
+  /// @param space (a const ref to) the IndexSpace object that specifies to this
+  /// space this object belongs
+  /// @param ord the index ordinal
+  /// @param proto_indices labels of proto indices (all must be unique,
+  /// i.e. duplicates are not allowed)
+  /// @param symmetric_proto_indices if true, proto_indices can be permuted at
+  /// will and will always be sorted
+  template <meta::integral I>
+  Index(const IndexSpace &space, std::optional<I> ord,
+        container::vector<Index> proto_indices,
+        bool symmetric_proto_indices = true)
+      : Index(space, ord) {
+    proto_indices_ = std::move(proto_indices);
+    symmetric_proto_indices_ = symmetric_proto_indices;
     canonicalize_proto_indices();
     check_for_duplicate_proto_indices();
-    check_nontmp_label();
   }
 
   /// @param label the index label, does not need to be unique, but must be
-  /// convertible into an IndexSpace (@sa IndexSpace::instance )
+  /// convertible into an IndexSpace (@sa IndexSpace::instance ); if @p label
+  /// is an rvalue reference will scavenge it for the label of the new index
+  /// @warning nonexplicit to make Index creation as ergonomic as possible, at
+  /// the cost of slightly increased danger
   template <typename String,
-            typename = std::enable_if_t<meta::is_basic_string_convertible_v<
-                std::remove_reference_t<String>>>>
+            typename = meta::EnableIfAllBasicStringConvertible<String>>
```

Comment:
We should turn this into a concept and then use a`requires` clause. That will yield much nicer error messages